templates/ts/client-browser.ts

```diff
@@ -80,40 +80,48 @@ export class {{name}}Client {
       ...this.options,
       ...options,
     };
-    const response = await fetch(`${this.serverUrl}/{{name}}`, {
-      ...mergedOptions,
-      headers: {
-        ...mergedOptions.headers,
-        'Content-Type': 'application/json',
-      },
-      body: JSON.stringify(body),
-      method: 'POST',
-    });
-    const isJSON = (response.headers.get('content-type') || '').startsWith('application/json');
+    let response: Response;
+    let responseBody: any;
+    let isJSON: boolean;
+    try {
+      response = await fetch(`${this.serverUrl}/{{name}}`, {
+        ...mergedOptions,
+        headers: {
+          ...mergedOptions.headers,
+          'Content-Type': 'application/json',
+        },
+        body: JSON.stringify(body),
+        method: 'POST',
+      });
+      isJSON = (response.headers.get('content-type') || '').startsWith('application/json');
+      if (isJSON && (response.status >= 200 && response.status < 300 || response.status === 400 || response.status === 500)) {
+        responseBody = await response.json();
+      }
+    } catch (err) {
+      throw new RequestError(err.message, err, "{{name}}", { serverUrl: this.serverUrl, ...this.options, ...options });
+    }
     if (response.status >= 200 && response.status < 300) {
       const validator = this.validators.{{{name}}};
-      const wrapped = { returns: isJSON ? await response.json() : undefined }; // wrapped for coersion
+      const wrapped = { returns: isJSON ? responseBody : undefined }; // wrapped for coersion
       if (!validator(wrapped)) {
         throw new ValidationError('Failed to validate response', validator.errors);
       }
       return wrapped.returns as {{{returnType}}};
     } else if (!isJSON) {
-      throw new Error(`${response.status} - ${response.statusText}`);
+      // fall through to throw
     } else if (response.status === 400) {
-      const body = await response.json();
-      if (body.name === 'ValidationError') {
-        throw new ValidationError(body.message, body.errors);
+      if (responseBody.name === 'ValidationError') {
+        throw new ValidationError(responseBody.message, responseBody.errors);
```

Comment:
Should we double check that responseBody is an object and has `message` and `errors`?

Reply:
In general we can probably JSON schema check the error responses, but I don't think it's likely we'll see a problem with them.

Reply:
Agree